### PR TITLE
Update documentation.

### PR DESCRIPTION
**Pull request by nomis52**

### CODE REVIEW

**Comment by peternewman** [comment]
Can you fix the Doxygen issues that Travis is catching please.

File: firmware/src/constants.h [hunk 1]
```diff
@@ -88,37 +88,48 @@
...
+   * @brief The command could not be completed due to a full memory buffer
+   */
   RC_BUFFER_FULL,
-  RC_BAD_PARAM,
-  RC_TX_ERROR,
-  RC_RX_TIMEOUT,
-  RC_RX_BCAST_RESPONSE,
-  RC_RX_INVALID_RESPONSE,
+  RC_BAD_PARAM,  //!< The command was malformed.
+  RC_TX_ERROR,  //!< The was an error during transceiver transmit.
```
Comment (peternewman) [inline firmware/src/constants.h]: SPaG There

File: firmware/src/constants.h [hunk 2]
```diff
@@ -88,37 +88,48 @@
...
+  RC_RDM_TIMEOUT,  //!< No response was received to RDM command.
```
Comment (peternewman) [inline firmware/src/constants.h]: SPaG to the

File: tests/doxygen/message-format.dox [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+ * The Host and Device communicate by exchanging messages in the form of
+ * frames. Each frame represents an operation or command. All communication is
+ * initiated by the Host and the Device sends a single reply per command.
+ *
+ * Messages sent from the Host to the Device are *Requests*, messages sent from
+ * the Device to the Host are *replies*.
+ *
+ * @section frame_endian Byte Ordering
+ *
+ * All multi-byte fields are sent little engian (LSB first)
```
Comment (peternewman) [inline tests/doxygen/message-format.dox]: SPaG endian

File: tests/doxygen/message-format.dox [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+ *
+ * <pre>
+ *    0                   1                   2                   3
+ *    0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+ *   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ *   \                         Data (if any)                         \
+ *   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ * </pre>
+ *
+ * @param data 0 or more bytes of data.
```
Comment (peternewman) [inline tests/doxygen/message-format.dox]: SPaG Data

File: tests/doxygen/message-format.dox [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+ *
+ * <pre>
+ *    0                   1                   2                   3
+ *    0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+ *   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ *   \                         Data (if any)                         \
+ *   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ * </pre>
+ *
+ * @param data 0 or more bytes of data, to match what was provided in the
```
Comment (peternewman) [inline tests/doxygen/message-format.dox]: SPaG Data

File: tests/doxygen/message-format.dox [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+ *
+ * The request contains no data.
+ *
+ * @subsubsection frame_cmd_getbreak_res Response Payload
+ *
+ * <pre>
+ *    0                   1
+ *    0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5
+ *   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ *   |         Break_time            |
```
Comment (peternewman) [inline tests/doxygen/message-format.dox]: SPaG Break_Time

File: tests/doxygen/message-format.dox [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+ *   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ * </pre>
+ *
+ * @param Break_time The current break time in microseconds.
```
Comment (peternewman) [inline tests/doxygen/message-format.dox]: SPaG Break_Time

File: tests/doxygen/message-format.dox [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+ *
+ * Sets the break time for outgoing DMX512 / RDM frames.
+ *
+ * @subsubsection frame_cmd_setbreak_req Request Payload
+ *
+ * <pre>
+ *    0                   1
+ *    0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5
+ *   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ *   |         Break_time            |
```
Comment (peternewman) [inline tests/doxygen/message-format.dox]: SPaG Break_Time

File: tests/doxygen/message-format.dox [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+ *   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ * </pre>
+ *
+ * @param Break_time The new break time. See Transceiver_SetBreakTime().
```
Comment (peternewman) [inline tests/doxygen/message-format.dox]: SPaG Break_Time

Comment (peternewman) [inline tests/doxygen/message-format.dox]: @see ?

File: tests/doxygen/message-format.dox [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+ *
+ * @subsection frame_cmd_unknown Unrecognised Commands
+ *
+ * If the device receives a command ID that is doesn't recognize it will return
+ * \ref RC_UNKNOWN.
+ *
+ * @section frame_transport Transport Considerations
+ * @subsection frame_transport_usb USB
+ *
+ * Comunication with the Ja Rule device happens over a custom endpoint.
```
Comment (peternewman) [inline tests/doxygen/message-format.dox]: SPaG Communication

File: tests/doxygen/message-format.dox [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+ *
+ * The maxPacketSize for the USB device is be 64 bytes. This is the largest
```
Comment (peternewman) [inline tests/doxygen/message-format.dox]: SPaG is 64

File: tests/doxygen/message-format.dox [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+ * bytes, we'll set the RX buffer size to 640 (64 * 10).
+ *
+ * On the Microchip RX side, the USB_DEVICE_EVENT_ENDPOINT_READ_COMPLETE event
+ * will be triggered if either:
+ *  - the host sends less than maxPacketSize data in any transaction
+ *  - the host sends maxPacketSize data, and the total size matches the rx
+ *    buffer size.
+ *
+ * libusb on Linux supports a Zero Length Packet (ZLP) flag, which will
+ * automatically sends a zero length packet if a transfer ends on a
```
Comment (peternewman) [inline tests/doxygen/message-format.dox]: SPaG send

File: tests/doxygen/message-format.dox [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+ * wMaxPacketSize boundary. Other host OS's don't seem to support this, so the
```
Comment (peternewman) [inline tests/doxygen/message-format.dox]: SPaG maxPacketSize unless its a Linux variable name

File: tests/doxygen/reference-design.dox [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+ * To build a basic DMX / RDM Transceiver module, you'll need:
+ *
+ *  - 1 x
+ *    <a href="http://datasheets.maximintegrated.com/en/ds/MAX1487-MAX491.pdf">
+ *    MAX481CPA</a>
+ *  - 2 x 562 &Omega; resistor
+ *  - 1 x 133 &Omega; resistor
+ *  - 1 x 0.1 &mu;F capacitor
+ *  - 1 x 5 pin XLR connector
+ *  - 1 x Breakout baord connector, or another means to attaching to J11 on the
```
Comment (peternewman) [inline tests/doxygen/reference-design.dox]: SPaG board

Comment (peternewman) [inline tests/doxygen/reference-design.dox]: SPaG means of

File: tests/doxygen/usb-logging.dox [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+ *
+ * [TOC]
+ *
+ * @section usb_Overview Overview
+ *
+ * Ja Rule devices provide a serial console over the USB port. This allows a
+ * user to connect using a <a
+ * href="https://en.wikipedia.org/wiki/Terminal_emulation">terminal
+ * emulator</a>, such as <a
+ * href="http://linux.die.net/man/1/minicom">minicom</a> (Unix) or Hyperterm
```
Comment (peternewman) [inline tests/doxygen/usb-logging.dox]: SPaG HyperTerminal

File: tests/doxygen/usb-logging.dox [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+ * (Windows).
```
Comment (peternewman) [inline tests/doxygen/usb-logging.dox]: Perhaps also add PuTTY, which works cross platform.

File: tests/doxygen/usb-logging.dox [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+ *
+ * The settings are 8N1, at 9600.
```
Comment (peternewman) [inline tests/doxygen/usb-logging.dox]: Add the word baud, possibly explain 8N1, or do we think people will have that specialised knowledge?

File: tests/doxygen/usb-logging.dox [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+ *
+ * The console provides different log levels. The +/- keys can be used to
+ * increase / decrease the log level.
+ *
+ * To log messages to the console, use SysLog_Message() and SysLog_Print().
+ * This should not be called from interrupt context.
```
Comment (peternewman) [inline tests/doxygen/usb-logging.dox]: SPaG from an

File: tests/mocks/Matchers.cpp [hunk 19]
```diff
@@ -22,6 +22,42 @@
...
+      : m_expected_data(expected_data),
+        m_expected_size(expected_size) {
+  }
+
+  bool MatchAndExplain(
+      ::testing::tuple<const void*, unsigned int> args,
+      testing::MatchResultListener* listener) const;
+
+  void DescribeTo(::std::ostream* os) const {
+    *os << "matches the data of size " << m_expected_size;
```
Comment (peternewman) [inline tests/mocks/Matchers.cpp]: Did this deal with the leading spaces?

File: tests/mocks/Matchers.cpp [hunk 20]
```diff
@@ -76,6 +112,42 @@ bool DataMatcher::InternalMatchAndExplain(
...
+    const uint8_t* expected_data,
+    unsigned int expected_size) {
+  return testing::MakeMatcher(new DataMatcher(expected_data, expected_size));
+}
+
+
+// PayloadMatcher
+// ----------------------------------------------------------------------------
+
+// A tuple matcher that checks a {iovec pointer, iovec size} tuple matches the
```
Comment (peternewman) [inline tests/mocks/Matchers.cpp]: Doxygenify

File: tests/mocks/Matchers.cpp [hunk 21]
```diff
@@ -22,6 +22,42 @@
 
 #include <gmock/gmock.h>
 
+// DataMatcher
+// ----------------------------------------------------------------------------
+
+/*
+ * @brief Check that a {pointer, length} tuple matches the expected data.
```
Comment (peternewman) [inline tests/mocks/Matchers.cpp]: Remove if it's in the .h

**Comment by peternewman** [comment]
Generally LGTM, apart from comments and Travis.

File: tests/doxygen/message-format.dox (hunk 9 above)
Comment (nomis52) [inline tests/doxygen/message-format.dox]: The name is automatically linked. Using @see puts it in a separate section which isn't what I want.

File: tests/doxygen/message-format.dox (hunk 13 above)
Comment (nomis52) [inline tests/doxygen/message-format.dox]: The name from the USB spec is wMaxPacketSize . I've added this in ( ) .

File: tests/doxygen/usb-logging.dox (hunk 17 above)
Comment (nomis52) [inline tests/doxygen/usb-logging.dox]: I think people will know.

File: tests/doxygen/usb-logging.dox (hunk 18 above)
Comment (nomis52) [inline tests/doxygen/usb-logging.dox]: changed to 'within'

File: tests/mocks/Matchers.cpp (hunk 19 above)
Comment (nomis52) [inline tests/mocks/Matchers.cpp]: As far as I remember it does.

**Comment by nomis52** [comment]
Hmm, so it looks like we need doxygen 1.8 on travis. Earlier versions don't understand the greek characters or #includes within example code.

**Comment by nomis52** [comment]
Peter: I've documented the rest of the important commands. Can you see if we can get a later version of doxygen on travis.

File: tests/mocks/Matchers.cpp (hunk 19 above)
Comment (peternewman) [inline tests/mocks/Matchers.cpp]: Okay great.

**Comment by peternewman** [comment]
Nice work getting a later Doxygen. LGTM.

Now you've got a later version of Doxygen, can you remove the comments, and uncomment the lines in the Doxyfile about "Commented out due to lack of support on Travis and causing spurious errors"

**Comment by nomis52** [comment]
Done.

**Pull request merged**